planner/core/rule_join_reorder_greedy.go

```diff
@@ -51,10 +55,62 @@ func (s *joinReorderGreedySolver) solve(joinNodePlans []LogicalPlan) (LogicalPla
 			cumCost: s.baseNodeCumCost(node),
 		})
 	}
+
+	for _, edge := range s.directedEdges {
+		nodeSet := s.directMap[edge[0]]
+		if nodeSet == nil {
+			nodeSet = make(map[LogicalPlan]struct{})
+			s.directMap[edge[0]] = nodeSet
+		}
+		nodeSet[edge[1]] = struct{}{}
+	}
+
+	isChanged := true
+	for isChanged {
+		isChanged = false
+		for _, set := range s.directMap {
+			oldSize := len(set)
+			newSet := make(map[LogicalPlan]struct{})
+			for k1 := range set {
+				for k2 := range s.directMap[k1] {
+					newSet[k2] = struct{}{}
+				}
+			}
+
+			for k2 := range newSet {
+				set[k2] = struct{}{}
+			}
+			newSize := len(set)
+			if oldSize != newSize {
+				isChanged = true
+			}
+		}
```

Comment:
This part actually propagates the direct edges, we can have a more clear comment here.